### PR TITLE
Support protocol fees on joins and exits (v2)

### PR DESCRIPTION
**Pull request by EndymionJkb**
Replaces #1626 - easier to tack this onto #1631 than trying to merge afterward.

The `_updatePostJoinExit` call sequences is not lovely, but it seemed even worse to special case the _afterJoinExit hook for initialization.

### CODE REVIEW

File: pkg/pool-weighted/contracts/test/MockWeightedPoolProtocolFees.sol [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../WeightedPoolProtocolFees.sol";
+
+contract MockWeightedPoolProtocolFees is WeightedPoolProtocolFees {
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/test/MockWeightedPoolProtocolFees.sol:20]: We don't actually use this in any tests it seems.

**Comment by EndymionJkb** [comment]
Right - it's preparation. Tests are in another PR (which needs to be
updated)

File: pkg/pool-weighted/contracts/test/MockWeightedPoolProtocolFees.sol (hunk 1 above)
Comment (EndymionJkb) [reply pkg/pool-weighted/contracts/test/MockWeightedPoolProtocolFees.sol:20]: I could remove it and put it in the separate test PR

File: pkg/pool-weighted/contracts/BaseWeightedPool.sol [hunk 2]
```diff
@@ -156,11 +156,18 @@ abstract contract BaseWeightedPool is BaseMinimalSwapInfoPool {
...
         uint256[] memory balanceDeltas,
-        uint256[] memory normalizedWeights
+        uint256[] memory normalizedWeights,
+        uint256 preJoinExitSupply,
+        uint256 postJoinExitSupply
     ) internal virtual {
         // solhint-disable-previous-line no-empty-blocks
     }
 
+    // Derived classes may call this to update state after a join or exit.
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/BaseWeightedPool.sol:166]:
```suggestion
    // Derived contracts may call this to update state after a join or exit.
```

File: pkg/pool-weighted/contracts/BaseWeightedPool.sol [hunk 3]
```diff
@@ -187,7 +194,8 @@ abstract contract BaseWeightedPool is BaseMinimalSwapInfoPool {
         // consistent in Pools with similar compositions but different number of tokens.
         uint256 bptAmountOut = Math.mul(invariantAfterJoin, amountsIn.length);
 
-        _afterJoinExit(true, new uint256[](amountsIn.length), amountsIn, normalizedWeights);
+        // Initialization is still a join, so we need to do post-join work.
+        _updatePostJoinExit(invariantAfterJoin);
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/BaseWeightedPool.sol:198]: Maybe mention why we're not calling `_afterJoinExit`

File: pkg/pool-weighted/contracts/BaseWeightedPool.sol [hunk 4]
```diff
@@ -209,14 +217,24 @@ abstract contract BaseWeightedPool is BaseMinimalSwapInfoPool {
         uint256[] memory normalizedWeights = _getNormalizedWeights();
 
         _beforeJoinExit(balances, normalizedWeights);
+
+        uint256 preJoinExitSupply = totalSupply();
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/BaseWeightedPool.sol:221]: You don't actually need to read it so early, as we haven't yet minted anything.

Also, we might save gas by passing this into `_doJoin`.

File: pkg/pool-weighted/contracts/WeightedPoolProtocolFees.sol [hunk 5]
```diff
@@ -64,23 +64,43 @@ abstract contract WeightedPoolProtocolFees is BaseWeightedPool, ProtocolFeeCache
             );
     }
 
-    function _afterJoinExit(
+    function _getJoinExitProtocolFees(
         bool isJoin,
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/WeightedPoolProtocolFees.sol:68]: Why don't need this any more, since we have both post and pre supply

Comment (TomAFrench) [reply pkg/pool-weighted/contracts/WeightedPoolProtocolFees.sol:68]: @EndymionJkb I think Nico means here that you could replace the usages of `isJoin` with `postJoinExitSupply > preJoinExitSupply`

Comment (EndymionJkb) [reply pkg/pool-weighted/contracts/WeightedPoolProtocolFees.sol:68]: Ah - that makes sense (now addressed with #1652 ). It looked like the *function* wasn't needed anymore somehow

**Review by nventuro — APPROVED** [review]

**Pull request merged**